src/controller/java/AndroidDeviceControllerWrapper.cpp

```diff
@@ -889,14 +910,59 @@ void AndroidDeviceControllerWrapper::OnICDRegistrationInfoRequired()
 void AndroidDeviceControllerWrapper::OnICDRegistrationComplete(chip::NodeId icdNodeId, uint32_t icdCounter)
 {
     chip::DeviceLayer::StackUnlock unlock;
+
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::app::ICDClientInfo clientInfo;
+    clientInfo.peer_node         = ScopedNodeId(icdNodeId, Controller()->GetFabricIndex());
+    clientInfo.monitored_subject = mAutoCommissioner.GetCommissioningParameters().GetICDMonitoredSubject().Value();
+    clientInfo.start_icd_counter = icdCounter;
+
+    ByteSpan symmetricKey = mAutoCommissioner.GetCommissioningParameters().GetICDSymmetricKey().Value();
+
+    err = mICDClientStorage.SetKey(clientInfo, symmetricKey);
+    if (err == CHIP_NO_ERROR)
+    {
+        err = mICDClientStorage.StoreEntry(clientInfo);
+    }
+
+    if (err == CHIP_NO_ERROR)
+    {
+        ChipLogProgress(Controller, "Saved ICD Symmetric key for " ChipLogFormatX64, ChipLogValueX64(icdNodeId));
+    }
+    else
+    {
+        mICDClientStorage.RemoveKey(clientInfo);
+        ChipLogError(Controller, "Failed to persist symmetric key for " ChipLogFormatX64 ": %s", ChipLogValueX64(icdNodeId),
+                     err.AsString());
+    }
+
+    mDeviceIsICD = true;
+
     JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
     jmethodID onICDRegistrationCompleteMethod;
-    CHIP_ERROR err = JniReferences::GetInstance().FindMethod(env, mJavaObjectRef, "onICDRegistrationComplete", "(JJ)V",
-                                                             &onICDRegistrationCompleteMethod);
+    jclass icdDeviceInfoClass         = nullptr;
+    jmethodID icdDeviceInfoStructCtor = nullptr;
+    jobject icdDeviceInfoObj          = nullptr;
+    jbyteArray jSymmetricKey          = nullptr;
+    err                               = JniReferences::GetInstance().FindMethod(env, mJavaObjectRef, "onICDRegistrationComplete",
```

Comment:
this err would override "err = mICDClientStorage.StoreEntry(clientInfo);", maybe we should use different variable called CHIP_ERROR methodErr here?

Reply:
Changed.